server/src/main/java/io/crate/planner/operators/InsertFromValues.java

```diff
@@ -642,11 +643,42 @@ private static ShardLocation getShardLocation(String indexName,
         AtomicInteger numRequests = new AtomicInteger(shardUpsertRequests.size());
         AtomicReference<Throwable> lastFailure = new AtomicReference<>(null);
 
+        Consumer<ShardUpsertRequest> countdown = request -> {
+            if (numRequests.decrementAndGet() == 0) {
+                Throwable throwable = lastFailure.get();
+                if (throwable == null) {
+                    result.complete(compressedResult);
+                } else {
+                    throwable = SQLExceptions.unwrap(throwable, t -> t instanceof RuntimeException);
+                    // we want to report duplicate key exceptions
+                    if (!SQLExceptions.isDocumentAlreadyExistsException(throwable) &&
+                            (partitionWasDeleted(throwable, request.index())
+                                    || partitionClosed(throwable, request.index())
+                                    || mixedArgumentTypesFailure(throwable))) {
+                        result.complete(compressedResult);
+                    } else {
+                        result.completeExceptionally(throwable);
+                    }
+                }
+            }
+        };
         for (ShardUpsertRequest request : shardUpsertRequests) {
-            String nodeId = state.routingTable()
-                .shardRoutingTable(request.shardId())
-                .primaryShard()
-                .currentNodeId();
+            String nodeId;
+            try {
+                nodeId = state.routingTable()
+                    .shardRoutingTable(request.shardId())
+                    .primaryShard()
+                    .currentNodeId();
+            } catch (IndexNotFoundException e) {
+                lastFailure.set(e);
+                if (!IndexParts.isPartitioned(request.index())) {
+                    synchronized (compressedResult) {
+                        compressedResult.markAsFailed(request.items());
+                    }
+                }
+                countdown.accept(request);
+                continue;
```

Comment:
just curious: why is the continue here needed?

Reply:
It should continue the loop, to have the same behavior as if the error happened in the actionListener.

Reply:
I may phrased it confusing: why is the continue needed to continue the loop as this is afaik the last statement inside the loop?

Reply:
Maybe the diff looks confusing? It's the first statement and the remainder of the loop body cannot work without a `nodeId`.

Reply:
ahh right, my fault! thx for explaining :)